nomt/src/beatree/ops/update/leaf_stage.rs

```diff
@@ -530,45 +515,45 @@ fn run_worker(
         leaf_updater.ingest(*key, value_change, overflow, delete_overflow);
     }
 
-    loop {
-        let res = leaf_updater.digest(&mut leaf_changes);
-
+    while let LeafDigestResult::NeedsMerge(cutoff) = leaf_updater.digest(&mut leaf_changes) {
         try_answer_left_neighbor(
             &mut pending_left_request,
             &mut worker_params,
             &mut leaf_changes,
+            has_finished_with_right_neighbor,
         );
 
-        if let LeafDigestResult::NeedsMerge(cutoff) = res {
-            if worker_params
-                .range
-                .high
-                .map_or(false, |high| cutoff >= high)
-            {
-                has_extended_range = true;
-                request_range_extension(&mut worker_params, &mut leaf_changes);
-            }
-
-            reset_leaf_base(
-                &bbn_index,
-                &leaf_cache,
-                &leaf_reader,
-                &mut leaf_changes,
-                &mut leaf_updater,
-                has_extended_range,
-                cutoff,
-            );
-
-            continue;
+        if worker_params
+            .range
+            .high
+            .map_or(false, |high| cutoff >= high)
+        {
+            has_extended_range = true;
+            request_range_extension(&mut worker_params, &mut leaf_changes);
         }
-        break;
+
+        reset_leaf_base(
+            &bbn_index,
+            &leaf_cache,
+            &leaf_reader,
+            &mut leaf_changes,
+            &mut leaf_updater,
+            has_extended_range,
+            cutoff,
+        );
     }
 
-    // answer any outstanding pending request
+    // Now we are safe to send over our right neighbor to the left one
+    // because we're sure to have finished interacting with it
+    has_finished_with_right_neighbor = true;
```

Comment:
this property is more clearly expressed as "has_finished_workload" or something like that, i.e. we've performed all inserts/splits/merges and therefore can safely give over our entire range to the left worker.